android-components/components/browser/menu2/src/test/java/mozilla/components/browser/menu2/ext/PopupWindowTest.kt

```diff
@@ -4,180 +4,388 @@
 
 package mozilla.components.browser.menu2.ext
 
+import android.graphics.Rect
 import android.view.Gravity
 import android.view.View
 import android.widget.PopupWindow
-import androidx.core.view.ViewCompat
+import androidx.recyclerview.widget.RecyclerView
 import androidx.test.ext.junit.runners.AndroidJUnit4
 import mozilla.components.browser.menu2.R
+import mozilla.components.browser.menu2.adapter.MenuCandidateListAdapter
 import mozilla.components.concept.menu.Orientation
-import mozilla.components.support.test.any
 import mozilla.components.support.test.robolectric.testContext
 import org.junit.Assert.assertEquals
 import org.junit.Before
 import org.junit.Test
 import org.junit.runner.RunWith
+import org.mockito.Mockito.any
+import org.mockito.Mockito.anyInt
 import org.mockito.Mockito.doAnswer
 import org.mockito.Mockito.doReturn
+import org.mockito.Mockito.mock
+import org.mockito.Mockito.never
 import org.mockito.Mockito.spy
 import org.mockito.Mockito.verify
-import org.robolectric.Shadows
-import org.robolectric.shadows.ShadowDisplay
+import org.mockito.Mockito.`when`
+import kotlin.math.roundToInt
+
+/**
+ * [PopupWindow] consts.
+ */
+private const val HALF_MENU_ITEM = 0.5
+
+/**
+ * [PopupWindow] UI components.
+ */
+private const val ANCHOR_ROOT_VIEW_HEIGHT = 1000
+private const val ANCHOR_ROOT_VIEW_WIDTH = 400
+private const val ITEM_HEIGHT = 50
+private const val CONTAINER_PADDING = 10
 
 @RunWith(AndroidJUnit4::class)
 class PopupWindowTest {
 
-    private lateinit var menuContentView: View
+    private lateinit var containerView: View
     private lateinit var anchor: View
     private lateinit var popupWindow: PopupWindow
 
     @Before
     fun setUp() {
-        menuContentView = createMockViewWith(x = 0, y = 0, false)
-        doReturn(90).`when`(menuContentView).measuredHeight
-        doReturn(30).`when`(menuContentView).measuredWidth
+        containerView = mock(View::class.java)
+        doReturn(createContainerRecyclerView()).`when`(containerView)
+            .findViewById<RecyclerView>(R.id.mozac_browser_menu_recyclerView)
 
         popupWindow = spy(PopupWindow())
-
-        // Makes the availableHeightToBottom bigger than the menuContentView
-        setScreenHeightAndWidth()
     }
 
     @Test
-    fun `WHEN displaying prefer down popup from top left THEN show popup down and to the right`() {
-        anchor = createMockViewWith(x = 0, y = 0, false)
-        popupWindow.displayPopup(menuContentView, anchor, Orientation.DOWN)
+    fun `WHEN recycler view has no adapter THEN show is never called`() {
```

Comment:
Good suggestion. I've updated to boolean as a starter - this may change based on feedback from on @rocketsroger around the displayPopup function error handling